src/main/java/reactor/kafka/receiver/observation/KafkaRecordReceiverContext.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2018-2023 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.kafka.receiver.observation;
+
+import java.nio.charset.StandardCharsets;
+
+import io.micrometer.observation.transport.ReceiverContext;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+
+/**
+ * {@link ReceiverContext} for {@link ConsumerRecord}s.
+ *
+ * @author Gary Russell
+ * @author Artem Bilan
+ * @since 1.4
+ *
+ */
+public class KafkaRecordReceiverContext extends ReceiverContext<ConsumerRecord<?, ?>> {
+
+    private final String clientId;
+
+    private final ConsumerRecord<?, ?> record;
+
+    public KafkaRecordReceiverContext(ConsumerRecord<?, ?> record, String clientId, String kafkaServers) {
+        super((carrier, key) -> {
+            Header header = carrier.headers().lastHeader(key);
+            if (header == null) {
+                return null;
+            }
+            return new String(header.value(), StandardCharsets.UTF_8);
+        });
+        setCarrier(record);
+        this.record = record;
```

Comment:
Since you are calling setCarrier to set the carrier and it has public get method, wouldn't it be better to simply use that instead of creating another reference with this.record ? Get source can then be like this:

```java
public String getSource(){
  return getCarrier().topic();
}
```

Reply:
Sure! We can do that, but it feels more like nit-picking than useful change.
In the end it is just pointer to the same object in the memory. Not sure what problem you see here...

Reply:
Oh it's not really a problem of a kind "it will leak/blow up". I just point out that it's duplicated content/reference which might confuse someone in future for a second.